### ISSUE TITLE
🐛 Bug Report — Runtime APIs — AbortSignal.timeout throws un-catchable async DOMException

### ISSUE THREAD

**Issue by aroman** [cloudflare/workerd#1020]
```ts
try {
  fetch("https://example.com", {
      signal: AbortSignal.timeout(1)
    }
} catch {}
```

expectation: when the promise is awaited, the fetch is aborted, and any error thrown is suppressed via the `catch {}` block

❌ reality: `✘ [ERROR] Uncaught (async) DOMException`

I assume this is a bug in workerd.

As a workaround, this works as expected:

try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 1);
    await fetch("https://example.com", {
    signal: controller.signal
  }
} catch {}

✅ no error message is reported, the error is fully suppressed by `catch {}`

wrangler 3.5.0

**Comment by kentonv** [comment]
Hi @aroman,

Can you provide a more complete reproduction? The code snippets you provided are not actually syntactically valid, and my best guess at what you meant appears to work correctly. It would help to see a whole worker script that exhibits the problem.

**Comment by aroman** [comment]
@kentonv yes; apologies for the sloppy snippets above. I have provided a complete, reliably reproducible demonstration of the issue below.

of note — I cannot seem to repro this with GET requests, it seems that I need a POST request to trigger the behavior. the repro example below demonstrates the behavior clearly. note that I can only reproduce this in miniflare, not running via `--remote`.


```js
// Run with: wrangler dev worker.js

export default {
  async fetch(req) {
    if (req.url.endsWith("/fail")) {
      await fail();
    } else if (req.url.endsWith("/fail-workaround")) {
      await failWorkaround();
    }
    return new Response("check the wrangler dev console");
  },
};

// Will log: ✘ [ERROR] Uncaught (async) DOMException
// Even though we're explicitly catching the error
// "caught error" will be logged
async function fail() {
  try {
    response = await fetch("https://example.com", {
      method: "POST",
      body: new FormData(),
      signal: AbortSignal.timeout(1),
    });
  } catch {
    console.log("caught error");
  }
}

// Will NOT cause the DOMException error to be logged
// "caught error" will be logged
async function failWorkaround() {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 1);
  try {
    response = await fetch("https://example.com", {
      method: "POST",
      body: new FormData(),
      signal: controller.signal,
    });
  } catch {
    console.log("caught error");
  }
  clearTimeout(timeout);
}
```

**Comment by kentonv** [comment]
Is it possible you're using an old version of Miniflare? Miniflare v2 tried to emulate Workers APIs in Node, whereas v3 actually uses `workerd` under the hood.

I haven't actually tested the above using raw `workerd`, but it does work correctly on cloudflareworkers.com and the underlying code is the same between them.

**Comment by aroman** [comment]
> Is it possible you're using an old version of Miniflare? Miniflare v2 tried to emulate Workers APIs in Node, whereas v3 actually uses `workerd` under the hood.
> 
> I haven't actually tested the above using raw `workerd`, but it does work correctly on cloudflareworkers.com and the underlying code is the same between them.

@kentonv 

I am using miniflare v3, which is bundled with the latest version of wrangler, `3.5.1`. Per the [release notes](https://github.com/cloudflare/workers-sdk/releases/tag/wrangler%403.5.1) for that version, it seems it is specifically using miniflare version `3.20230814.1`.

As I mentioned in the OP, this behavior is **only reproducible running via miniflare**, and NOT running on cloudflare's hosted workers environment (as in via `--remote`).

**Comment by aroman** [comment]
Thanks for your efforts to repro, @mrbbot! Good to know I'm not going crazy 😅

Is the `--verbose` flag on by default for `wrangler dev`? I've never invoked `workerd` directly, only via `wrangler`.

**Comment by kentonv** [comment]
OK, I may have misunderstood what I was looking for when I tried this. So the exception isn't actually "un-catchable", it's just that there's a spurious log message?

**Comment by aroman** [comment]
> OK, I may have misunderstood what I was looking for when I tried this. So the exception isn't actually "un-catchable", it's just that there's a spurious log message?

@kentonv yes, that is correct. _even though_ the error is caught (i.e. the catch block runs), the message "✘ [ERROR] Uncaught (async) DOMException" still appears. i am not sure if this will actually bubble up as an actual uncaught promise rejection (as the log suggests), or the log is totally inaccurate, but either way, it's a bug and confusing.

**Comment by kentonv** [comment]
@jasnell Apparently `AbortSignal.timeout()` generates a bogus unhandled rejection event, whereas using `setTimeout()` and `controller.abort()` to implement a timeout does not. Any idea why?

**Comment by kentonv** [comment]
Err actually I'm not sure if it's an unhandled rejection event or just something _like_ an unhandled rejection event, but either way...

**Comment by jasnell** [comment]
Sorry just spotted this. will investigate

**Comment by jasnell** [comment]
Ok, @kentonv the issue is not specific to `AbortSignal.timeout()` here, you'll see that in the case of calling `controller.abort('foo')` within the `setTimeout()` the error is also logged then. The reason is that within the `AbortSignal` we are using a `kj::Canceler` to actually cancel the underlying operation on the fetch which means we're creating another propagation path for the error that is being logged. That path operates separately from the try/catch that is in scope here so would be logged separately. I'll investigate further to see how we can suppress it.

**Comment by jasnell** [comment]
Actually, let me take that back... it *is `AbortSignal.timeout` in terms of the way that it builds the error internally. Specifically, `AbortSignal.timeout` reports the reason for the disconnect using a kj::Exception converted to a JS exception, which gets converted back into a `kj::Exception` (this part can be cleaned up a bit). The code is set up to ignore logging for kj::Exception's that are marked as `DISCONNECTION` errors but in this case that's not reported. It ends up being no different than if `AbortController.abort(...)` is called with a user-supplied reason. Should be fairly straightforward to clean up

**Comment by jasnell** [comment]
@aroman ... I have a PR that fixes the specific issue when using `AbortSignal.timeout()`, but I have a question for you.

Here's the set up:

When we have a `fetch` request that that has a request body payload (e.g. `POST`, `PUT` etc) to be transmitted to the server, there are actually *two* distinct async operations involved: The promise that is pending for the `Response` (which I'll call `P1`) and the promise that is pending for the actual transmission of the request body payload (which I'll call `P2`). The `AbortSignal` actually cancels *both* of these by passing along the `reason` that is given to the `AbortSignal`. This `reason` is typically an `Error` but can be any JavaScript value.

`P1`, the promise for the `Response`, is the thing that you are actually awaiting for when you call `await fetch(...)`. That promise is resolved *independently* of the transmission of the request body payload (`P2`). Your `try/catch` in your reproduction is ONLY handling the rejection of `P1`. The other promise (`P2`) is handled internally and is NOT something user code can handle directly. The "unhandled exception" that you are seeing here is from the rejection of `P2`.

We currently have code in place that when you abort a `fetch` using an `AbortController`, and you do *not* pass in a reason explicitly (calling `abortController.abort()` with no arguments), the rejection of `P2` is handled *silently* within the runtime. That is, you will *NOT* see the unhandled exception notice for `P2`. When you call `abortController.abort(reason)`, where `reason` is any JavaScript value, you *WILL* see the unhandled exception notice for `P2`. Using `AbortSignal.timeout(n)` is currently equivalent to `abortController.abort(reason)`. The PR that I have open in #1177 causes the `AbortSignal.timeout(n)` to also be handled silently, addressing the specific issue that you have reported here.

Here comes the question for you: Even after #1177 lands, calling `abortController.abort(reason)` will cause the unhandled exception error to be reported. This is a slightly different case than what you reported and it would be helpful to understand what you, as a user, would expect in this case.

Consider the following example. We are assuming that the server side of this request is duplexing their response... that is, they are sending the response headers *before* the client has completely finished transmitting the request payload. This would mean that the `await fetch` promise resolves *before* the request payload has finished. The abort in this case will end up stopping the transmission of the payload abruptly after the fetch promise has already resolved.

```
  const enc = new TextEncoder();
  const rs = new ReadableStream({
    type: 'bytes',
    async pull(c) {
      await scheduler.wait(100);
      // stream never ends...
      c.enqueue(enc.encode('hello'));
    }
  });
  const ac = new AbortController();
  try {
    const response = await fetch("https://example.com/", {
      method: "POST",
      body: rs,
      signal: ac.signal,
    });
    // we've received the response headers but the body payload is still running... let's
    // use the AbortController to interrupt
   ac.abort(new Error('canceled'));
  } catch {
    // never runs because the fetch promise itself resolved just fine....
    console.log("caught error");
  }
```

In this scenario, because we passed an explicit reason to `ac.abort(...)`, the runtime will report in the logs that an unhandled exception occurred. This exception is not catchable by user code. If we do not log the exception, then you might not have any way of knowing *why* the request payload body was never transmitted. You might never actually know whether it was or was not actually transmitted at all.

PR #1177 *does* fix the problem you originally raised but it *does not* address this additional concern about `ac.abort(new Error('canceled')` in this case. The question to you is, would the unhandled exception log in such a case *also* be problematic for you? Would you prefer that we not emit any unhandled exception messages in this case? What would be your expectation here?
